pkg/api/operators/apis/experiment/v1alpha2/experiment_types.go

```diff
@@ -236,12 +236,64 @@ type Operation struct {
 	Parameters    []ParameterSpec `json:"parameters,omitempty"`
 }
 
-// Spec for metrics collectors. For v1alpha2 we will keep metrics collection as
-// cron jobs. This can be replaced by a push-model when the metadata client lib
-// is ready.
 type MetricsCollectorSpec struct {
-	Retain     bool       `json:"retain,omitempty"`
+	// Deprecated Retain
+	Retain bool `json:"retain,omitempty"`
+	// Deprecated GoTemplate
 	GoTemplate GoTemplate `json:"goTemplate,omitempty"`
+
+	Source    *SourceSpec    `json:"source,omitempty"`
+	Collector *CollectorSpec `json:"collector,omitempty"`
+}
+
+type SourceSpec struct {
+	// Model-train source code can expose metrics by http, such as HTTP endpoint in
+	// prometheus metric format
+	HttpGet *v1.HTTPGetAction `json:"httpGet,omitempty"`
+	// During training model, metrics may be persisted into local file in source
+	// code, such as tfEvent use case
+	FileSystemPath *FileSystemPath `json:"fileSystemPath,omitempty"`
+	// Default metric output format is {"metric": "<metric_name>",
+	// "value": <int_or_float>, "epoch": <int>, "step": <int>}, but if the output doesn't
+	// follow default format, please extend it here
+	Filter *FilterSpec `json:"filter,omitempty"`
+}
+
+type FilterSpec struct {
+	// When the metrics output follows format as this field specified, metricsCollector
+	// collects it and reports to metrics server, it can be "<metric_name>: <float>" or else
+	MetricsFormat []string `json:"metricsFormat,omitempty"`
```

Comment:
MetricsFormat is not clear to me. Does user need to specify objective metric name again here?

Reply:
for example, we can make MetricsFormat  like `["{{metricName}} := {{metricValue}}", "{{metricName}} is {{metricValue}}"]`, and the experiment objectivemetrics is `loss`, and additional metrics `accuracy`, metrics will run logic like
```
for line of outputLines:
  for metric in ["loss", "accuracy"]:
    for format in MetricsFormat:
      mformat = convert_metrics(metric, format) // use metric replace {{metricName}}
      value = parse(line, mformat) // try to match the output like "loss  := $something" or "loss is $something", and return $something
      if !value:
         continue
      report_metrics(metric, value)
```

Reply:
So, you mean to say that `metricName`, `metricValue`  are keywords to be used in Spec for specifying metricFormat.  But the comment says - "<metric_name>: <float>"

Reply:
the comment is about default format with json format, which keeps same as floydhub.
in `MetricsFormat`, we can use {{metricName}} and {{metricValue}} as a placeholder.
anyway, maybe we can talk more about what the placeholder should looks like in another thread.